src/stratis/timer.rs

```diff
@@ -67,18 +73,107 @@ async fn check_pool_and_fs(
     }
 }
 
+pub struct TokioEventHandler(watch::Sender<Option<notify::Result<notify::Event>>>);
+
+impl EventHandler for TokioEventHandler {
+    fn handle_event(&mut self, event: notify::Result<notify::Event>) {
+        match event {
+            Ok(e) => {
+                if let notify::EventKind::Modify(_) = e.kind {
+                    if let Err(e) = self.0.send(Some(Ok(e))) {
+                        warn!("Failed to send inotify event to processing thread: {e}");
+                    }
+                }
+            }
+            Err(e) => {
+                if let Err(e) = self.0.send(Some(Err(e))) {
+                    warn!("Failed to send inotify event to processing thread: {e}");
+                }
+            }
+        }
+    }
+}
+
+/// Periodically refresh the keyring to ensure the keys are never garbage collected.
+async fn refresh_keyring() {
+    async fn process_timeout() -> StratisResult<u64> {
+        let mut file = File::open("/proc/sys/kernel/keys/persistent_keyring_expiry").await?;
+        let mut number = String::new();
+        file.read_to_string(&mut number).await?;
+        let trimmed = number.trim();
+        trimmed
+            .parse::<u64>()
+            .map_err(|e| StratisError::Msg(e.to_string()))
+    }
+
+    async fn process_checks() -> StratisResult<()> {
+        get_persistent_keyring().map(|_| ())
+    }
+
+    fn initialize_inotify(
+        tx: watch::Sender<Option<notify::Result<notify::Event>>>,
+    ) -> notify::Result<RecommendedWatcher> {
+        let mut watcher = notify::recommended_watcher(TokioEventHandler(tx))?;
+        watcher.watch(
+            Path::new(KEY_EXPIRY_PATH),
+            notify::RecursiveMode::NonRecursive,
+        )?;
+        Ok(watcher)
+    }
+
+    let (tx, mut rx) = watch::channel::<Option<notify::Result<notify::Event>>>(None);
+    let _watcher = match initialize_inotify(tx) {
+        Ok(w) => Some(w),
+        Err(e) => {
+            warn!("Failed to initialize inotify: {e}; keyring expiry may not be accurately tracked in stratisd");
+            None
+        }
+    };
+
+    loop {
+        trace!("Starting timed keyring refresh");
+        let timeout = match process_timeout().await {
+            Ok(t) => t,
+            Err(e) => {
+                warn!("Failed to get timeout for keyring refresh: {e}");
+                continue;
+            }
+        };
+        if let Err(e) = process_checks().await {
+            warn!("Failed to refresh keyring: {e}");
+        }
+        trace!("Timed keyring refresh finished");
+        let timeout_adjusted = timeout.saturating_sub(3600);
```

Comment:
I think we should avoid our timeout actually going to 0, otherwise the task will always be wanting to run. If the timeout is so short that it is better not to refresh at all, we can log a warning. But we should only do it once, on the notify on the ```persistent_keyring_expiry``` file.

Reply:
The other option I wanted to propose is that we use the process keyring instead. This means that we will never preserve any keys across restarts of stratisd but we will not have to refresh. Is that preferable?

Reply:
We should discuss in standup. Any use to using the process keyring generally, but attaching to the persistent keyring when stratisd shuts down?